terraform/node_provider.go

```diff
@@ -1,11 +1,114 @@
 type NodeApplyableProvider struct {
 	*NodeAbstractProvider
 }
 
-// GraphNodeEvalable
-func (n *NodeApplyableProvider) EvalTree() EvalNode {
-	return ProviderEvalTree(n, n.ProviderConfig())
+var (
+	_ GraphNodeExecutable = (*NodeApplyableProvider)(nil)
+)
+
+// GraphNodeExecutable
+func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) error {
+	_, err := ctx.InitProvider(n.Addr)
+	if err != nil {
+		return err
+	}
+	provider, _, err := GetProvider(ctx, n.Addr)
+	if err != nil {
+		return err
+	}
+
+	switch op {
+	case walkValidate:
+		//Validate
+		return n.ValidateProvider(ctx, provider)
+	case walkRefresh, walkPlan, walkApply, walkDestroy:
+		// EvalConfigProvider
+		return n.ConfigureProvider(ctx, provider, false)
+	case walkImport:
+		return n.ConfigureProvider(ctx, provider, true)
+	}
+	return nil
+}
+
+func (n *NodeApplyableProvider) ValidateProvider(ctx EvalContext, provider providers.Interface) error {
+	var diags tfdiags.Diagnostics
+
+	configBody := buildProviderConfig(ctx, n.Addr, n.ProviderConfig())
+
+	resp := provider.GetSchema()
+	diags = diags.Append(resp.Diagnostics)
+	if diags.HasErrors() {
+		return diags.ErrWithWarnings()
+	}
+
+	configSchema := resp.Provider.Block
+	if configSchema == nil {
+		// Should never happen in real code, but often comes up in tests where
+		// mock schemas are being used that tend to be incomplete.
+		log.Printf("[WARN] ValidateProvider: no config schema is available for %s, so using empty schema", n.Addr)
+		configSchema = &configschema.Block{}
+	}
+
+	configVal, configBody, evalDiags := ctx.EvaluateBlock(configBody, configSchema, nil, EvalDataForNoInstanceKey)
+	diags = diags.Append(evalDiags)
+	if evalDiags.HasErrors() {
+		return diags.ErrWithWarnings()
+	}
+
+	req := providers.PrepareProviderConfigRequest{
+		Config: configVal,
+	}
+
+	validateResp := provider.PrepareProviderConfig(req)
+	diags = diags.Append(validateResp.Diagnostics)
+
+	return diags.ErrWithWarnings()
+}
+
+func (n *NodeApplyableProvider) ConfigureProvider(ctx EvalContext, provider providers.Interface, VerifyConfigIsKnown bool) error {
+	var diags tfdiags.Diagnostics
+	config := n.ProviderConfig()
+
+	configBody := buildProviderConfig(ctx, n.Addr, config)
+
+	resp := provider.GetSchema()
+	diags = diags.Append(resp.Diagnostics)
+	if diags.HasErrors() {
+		return diags.ErrWithWarnings()
+	}
+
+	configSchema := resp.Provider.Block
+	configVal, configBody, evalDiags := ctx.EvaluateBlock(configBody, configSchema, nil, EvalDataForNoInstanceKey)
+	diags = diags.Append(evalDiags)
+	if evalDiags.HasErrors() {
+		return diags.ErrWithWarnings()
+	}
+
+	if VerifyConfigIsKnown && !configVal.IsWhollyKnown() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid provider configuration",
+			Detail:   fmt.Sprintf("The configuration for %s depends on values that cannot be determined until apply.", n.Addr),
+			Subject:  &config.DeclRange,
+		})
+		return diags.ErrWithWarnings()
```

Comment:
You can't easily tell, but I had to change the return types on this function from diags.NonFatalError to diags.ErrWithWarnings(). It turns out that EvalSequence, which was handling ~all this in ProviderEvalTree, handled exit errors differently than Eval (eval_sequence.go and eval.go). 

All errors from `Execute`s are handling consistently, so some modifications to return values are likely to come up again.